Core/clim-basic/clipboard.lisp

```diff
@@ -0,0 +1,76 @@
+(in-package :climi)
+
+;;;
+;;;  Functions dealing with copying to the clipboard.
+;;;
+
+(deftype representation-type-name () '(member :string :html :image))
+
+(defgeneric copy-to-clipboard-with-port (port sheet clipboard-p object presentation-type)
+  (:documentation "Method to be implemented by backends."))
+
+(defmethod copy-to-clipboard-with-port ((port clim:port) (sheet clim:sheet) clipboard-p object presentation-type)
+  "Fallback implementation if the implementation does not support a clipboard."
+  nil)
+
+(defun copy-to-clipboard (sheet object &key presentation-type)
+  "Copy OBJECT to the clipboard.
+SHEET is the owner of the clipboard, and it is not guaranteed that the
+content of the clipboard will be available after the sheet has been
+removed."
+  (copy-to-clipboard-with-port (port sheet) sheet t object presentation-type))
+
+(defun copy-to-selection (sheet object &key presentation-type)
+  "Copy OBJECT to the selection.
+SHEET is the owner of the selection, and it is not guaranteed that the
+content of the selection will be available after the sheet has been
+removed."
+  (copy-to-clipboard-with-port (port sheet) sheet nil object presentation-type))
```

Comment:
what does it mean to copy to selection? how is it different from copying to the clipboard? is it just that "copy-to-clipboard" is copy and "copy-to-selection" is paste?

Reply:
No. As explained in the overview document (which has not been merged into this change yet), there are two separate concepts: The *selection* is the the highlighted text. That's the text that is pasted when you middle-click in most X11 applications.

The *clipboard* is a separate thing. It's where the information that is pasted using Control-V in most applications copy from. When typing Control-C, the content of the selection is copied to the clipboard.

Thus, when text is highlighted, `COPY-TO-SELECTION` is called to inform the system that something has been selected. In a later commit, a separate command will be added that implements the behaviour of Control-C in most applications, and it checks if some text is selected, extracts it, and calls `COPY-TO-CLIPBOARD` on it.

Reply:
please split this new protocol between selection and clipboard modules. fake unification (with clipboard-p) buys us nothing but still imposes extra cognitive effort on backend implementers (even if selection doesn't make sense there).

also remove extra functions (i.e instead of clear-clipboard and clear-clipboard-with-port make one generic function clear-sheet-clipboard accepting both port and a sheet (that's how i.e set-sheet-cursor is implemented, so it would be consistent).

So on example of clear-sheet-*:

Remove clear-clipboard-with-port and change clear-clipboard / clear-selection to generic functions with arguments (port sheet). moreover move clear-selection to selection module.

Reply:
Changed according to recommendation.